mailbox/store/src/main/java/org/apache/james/mailbox/store/search/ListeningMessageSearchIndex.java

```diff
@@ -68,45 +69,51 @@ public boolean isHandling(Event event) {
      * something relevant is received
      */
     @Override
-    public void event(Event event) throws Exception {
-        handleMailboxEvent(event,
-            sessionProvider.createSystemSession(event.getUsername()),
-            (MailboxEvent) event);
+    public Mono<Void> reactiveEvent(Event event) {
+        try {
+            return handleMailboxEvent(event,
+                sessionProvider.createSystemSession(event.getUsername()),
+                (MailboxEvent) event);
+        } catch (Exception e) {
+            return Mono.error(e);
+        }
     }
 
-    private void handleMailboxEvent(Event event, MailboxSession session, MailboxEvent mailboxEvent) throws Exception {
+    private Mono<Void> handleMailboxEvent(Event event, MailboxSession session, MailboxEvent mailboxEvent) throws Exception {
         MailboxId mailboxId = mailboxEvent.getMailboxId();
 
         if (event instanceof Added) {
             Mailbox mailbox = factory.getMailboxMapper(session).findMailboxById(mailboxId);
-            handleAdded(session, mailbox, (Added) event);
+            return handleAdded(session, mailbox, (Added) event);
         } else if (event instanceof Expunged) {
             Mailbox mailbox = factory.getMailboxMapper(session).findMailboxById(mailboxId);
             Expunged expunged = (Expunged) event;
-            delete(session, mailbox, expunged.getUids());
+            return delete(session, mailbox, expunged.getUids());
         } else if (event instanceof FlagsUpdated) {
             Mailbox mailbox = factory.getMailboxMapper(session).findMailboxById(mailboxId);
             FlagsUpdated flagsUpdated = (FlagsUpdated) event;
-            update(session, mailbox, flagsUpdated.getUpdatedFlags());
+            return update(session, mailbox, flagsUpdated.getUpdatedFlags());
         } else if (event instanceof MailboxDeletion) {
-            deleteAll(session, mailboxId);
+            return deleteAll(session, mailboxId);
         }
+        return Mono.empty();
     }
 
-    private void handleAdded(MailboxSession session, Mailbox mailbox, Added added) {
-        MessageRange.toRanges(added.getUids())
-            .stream()
-            .flatMap(range -> retrieveMailboxMessages(session, mailbox, range))
-            .forEach(Throwing.<MailboxMessage>consumer(mailboxMessage -> add(session, mailbox, mailboxMessage)).sneakyThrow());
+    private Mono<Void> handleAdded(MailboxSession session, Mailbox mailbox, Added added) {
+        return Flux.fromIterable(MessageRange.toRanges(added.getUids()))
+            .concatMap(range -> retrieveMailboxMessages(session, mailbox, range))
+            .concatMap(mailboxMessage -> add(session, mailbox, mailboxMessage))
```

Comment:
Why does it have to be done one by one  and in order ?

Reply:
It's more "async background process shouldn't obliterrate backend resource usage".

The stream version wasn't parrallel. I'm unconfortable with unbounded resource usage. I don't want to alter that behaviour.